Create usaco-1230.mdx

EulerPascal404 opened the pull request:
Adding the march silver #1 problem - it is a good application of functional graphs. I will add it to functional graphs module.

_If any of the below doesn't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which includes the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making them.

EulerPascal404 commented:
What is the problem, would anyone mind helping me with this?

envyaims commented on solutions/usaco-1230.mdx:
```diff
@@ -0,0 +1,78 @@
+---
+id: usaco-1230
+source: USACO Silver 2022 Open
+title: Redistributing Gifts
```
wrong problem name lol

envyaims reviewed (changes requested):
bruh you just copied the official solution -_-

http://www.usaco.org/current/data/sol_prob1_silver_open22.html

review this from https://usaco.guide/general/contributing
![image](https://user-images.githubusercontent.com/79723780/169675153-67d5fb04-ddd3-4118-b4a5-17e945634f0f.png)

envyaims commented:
> Adding the march silver #1 problem - it is a good application of functional graphs. I will add it to functional graphs module.

isnt it already there?

EulerPascal404 commented:
Hey envyaims!

I am a new contributor on the site, so I didn't realize the editorial solutions were not supposed to be added directly to the guide. I have changed the commit with my solutions(and added comments for it to be more helpful). As for the functional graphs module, I meant that the solution was not linked there - but I see that I was wrong, it was just linked directly to the editorial.

Thanks!

SansPapyrus683 commented on solutions/usaco-1230.mdx:
```diff
@@ -0,0 +1,92 @@
+
+<CPPSection>
+
+This code is similar to the official editorial, but more intuitive and with comments.
+	
```
```suggestion
```
not really needed

SansPapyrus683 commented on solutions/usaco-1230.mdx:
````diff
@@ -0,0 +1,92 @@
+This code is similar to the official editorial, but more intuitive and with comments.
+	
+```cpp
+#include <iostream>
````
```suggestion
```
you're using `bits` already

SansPapyrus683 commented on solutions/usaco-1230.mdx:
```diff
@@ -0,0 +1,92 @@
+#include <iostream>
+#include <bits/stdc++.h>
+
+#define int int64_t
```
dear god please no `#define`

envyaims replied:
using ll = long long 😎

SansPapyrus683 commented on solutions/usaco-1230.mdx:
```diff
@@ -0,0 +1,92 @@
+        y = a[y];
+    }
+    mark(y);
+    return min_v;
```
tabs

SansPapyrus683 commented on solutions/usaco-1230.mdx:
```diff
@@ -0,0 +1,92 @@
+    return min_v;
+}
+
+int32_t main()
+{
```
either put the opening brace on the same line or a new line all the time

SansPapyrus683 commented on solutions/usaco-1230.mdx:
```diff
@@ -0,0 +1,92 @@
+
+int32_t main()
+{
+    int n, ans = 0;
```
move `ans` down where you actually use it

SansPapyrus683 commented on solutions/usaco-1230.mdx:
```diff
@@ -0,0 +1,92 @@
+        if (!visited[i]) ans -= min_in_cycle(i);
+    }
+
+    cout << ans << "\n";
+    return 0;
```
```suggestion
    cout << ans << endl;
```

SansPapyrus683 commented on solutions/usaco-1230.mdx:
````diff
@@ -0,0 +1,92 @@
+
+```
+</CPPSection>
+
+<PySection>
+</PySection>
+
+<JavaSection>
+
+</JavaSection>
+
````
```suggestion
```
what

SansPapyrus683 commented on solutions/usaco-1230.mdx:
```diff
@@ -0,0 +1,92 @@
+
+vector<int> a,v; //cow i wants to visit cow a[i] and gets v[i] points
+vector<vector<int>> graph; //graph[i] stores the cows that want to go to farm i
+vector<bool> visited; //done
```
> done

what this mean

envyaims commented on solutions/usaco-1230.mdx:
```diff
@@ -0,0 +1,92 @@
+        y = a[y];
+        z = a[a[z]];
+    }
+    int min_v = v[y];
```
some comments throughout here would be nice

and better variable names...

envyaims commented on solutions/usaco-1230.mdx:
```diff
@@ -0,0 +1,92 @@
+	for (int c : graph[y]) mark(c);
+}
+
+//Applies Floyd's algorithm
```
space out these comments like:

```
// text
```

envyaims reviewed (changes requested):
overall some more comments would be nice

otherwise it's fine

EulerPascal404 commented:
@envyaims @SansPapyrus683 I think I have made all the requested changes + added more comments. Please merge if it's fine, or if there are other errors, please let me know.

envyaims commented:
Fix comment spacing for all other comments and use tabs.

EulerPascal404 commented:
I did not know what you meant when adding the tabbing(since it was there in my local IDE). I have fixed it and made the comments uniform.

envyaims commented on solutions/usaco-1230.mdx:
```diff
@@ -0,0 +1,83 @@
+    	for(int i = 0; i < n; i++){
+        	cin >> a[i] >> v[i];
+        	a[i]--;
+        	graph[a[i]].push_back(i); //since cow i wants to visit farm a[i]
```
```suggestion
        	graph[a[i]].push_back(i); // cow i wants to visit farm a[i]
```

envyaims commented on solutions/usaco-1230.mdx:
```diff
@@ -0,0 +1,83 @@
+int min_in_cycle(int curr){
+	int y = a[curr];
+	int z = a[y];
+	while(y != z){
```
```suggestion
	while (y != z) {
```

keep your spacings consistent

envyaims commented on solutions/usaco-1230.mdx:
```diff
@@ -0,0 +1,83 @@
+	y = a[y];
+
+	// traverse the cycle to find the cow with the minimum v_i value
+	while(y != z){
```
```suggestion
	while (y != z) {
```

envyaims commented on solutions/usaco-1230.mdx:
```diff
@@ -0,0 +1,83 @@
+    	graph.resize(n);
+
+    	ll ans = 0;
+    	for(int i = 0; i < n; i++){
```
```suggestion
    	for (int i = 0; i < n; i++) {
```

envyaims commented on solutions/usaco-1230.mdx:
```diff
@@ -0,0 +1,83 @@
+
+    	// from each cycle, you can visit all except 1 cow. We make this cow the one
+    	// with the lowest v_i. We subtract this from our total answer.
+    	for (int i = 0; i < n; i++){
```
```suggestion
    	for (int i = 0; i < n; i++) {
```

EulerPascal404 commented:
Committed all the new suggestions.

envyaims commented on solutions/usaco-1230.mdx:
```diff
@@ -0,0 +1,83 @@
+
+typedef long long ll;
+
+vector<int> a,v; // cow i wants to visit cow a[i] and gets v[i] points
```
```suggestion
vector<int> a, v; // cow i wants to visit cow a[i] and gets v[i] points
```

envyaims reviewed (approved):
great!

EulerPascal404 commented:
I committed the change(although it seems rather like splitting hairs).

SansPapyrus683 commented:
it's best to space out your operators so it's more readable

EulerPascal404 commented:
> it's best to space out your operators so it's more readable

Got it.

SansPapyrus683 commented on solutions/usaco-1230.mdx:
```diff
@@ -0,0 +1,83 @@
+
+// marks y and the other cows in its other cycle as visited
+void mark(int y){
+	if (visited[y]) return;
+	visited[y] = true;
+	for (int c : graph[y]) mark(c);
```
braces

EulerPascal404 commented:
Done.

SansPapyrus683 commented:
wait for deployment then i'll merge

EulerPascal404 commented:
Alright, thank you!
